### PR TITLE
KAFKA-16154: Broker returns offset for LATEST_TIERED_TIMESTAMP

### PR DESCRIPTION
**Pull request by brandboat**
This pr support EarliestLocalSpec LatestTierSpec in GetOffsetShell, and add integration tests.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/admin/OffsetSpec.java [hunk 1]
```diff
@@ -70,4 +72,21 @@ public static OffsetSpec maxTimestamp() {
         return new MaxTimestampSpec();
     }
 
+    /**
+     * Used to retrieve the offset with the local log start offset,
+     * log start offset is the offset of a log above which reads are guaranteed to be served
+     * from the disk of the leader broker, when Tiered Storage is not enabled, it behaves the same
+     * as the earliest timestamp
```
Comment (showuon) [inline clients/src/main/java/org/apache/kafka/clients/admin/OffsetSpec.java]: nit: Used to retrieve ~~the offset with~~ the local log start offset[.] 
[Local] log start offset is the offset of a log above which reads are guaranteed to be served from the disk of the leader broker.

Note: When tiered storage is not enabled, it behaves the same as retrieving the earliest timestamp offset.

File: clients/src/main/java/org/apache/kafka/clients/admin/OffsetSpec.java [hunk 2]
```diff
@@ -70,4 +72,21 @@ public static OffsetSpec maxTimestamp() {
...
+     */
+    public static OffsetSpec earliestLocalSpec() {
+        return new EarliestLocalSpec();
+    }
+
+    /**
+     * Used to retrieve the offset with the highest offset of data stored in remote storage,
+     * and when Tiered Storage is not enabled, we won't return any offset (i.e. Unknown offset)
```
Comment (showuon) [inline clients/src/main/java/org/apache/kafka/clients/admin/OffsetSpec.java]: nit: 

Used to retrieve ~~the offset with~~ the highest offset of data stored in remote storage[.]

Note: When tiered storage is not enabled, we ~~won't return any offset (i.e.~~  [will return] unknown offset.

**Review by showuon — COMMENTED** [review]
Thanks for the PR. Left some comments.

File: clients/src/main/java/org/apache/kafka/clients/admin/OffsetSpec.java (hunk 2 above)
Comment (brandboat) [reply clients/src/main/java/org/apache/kafka/clients/admin/OffsetSpec.java]: Thanks for the comments, already fixed them in the latest commit.

**Review by showuon — COMMENTED** [review]
Overall LGTM! We need integration tests for it. We should add some test like `DeleteTopicTest` using `ExpectListOffsetsAction`. I expected it should work like this (I didn't try):

```
// send records to partition 0
                .expectSegmentToBeOffloaded(broker0, topicA, p0, 0, new KeyValueSpec("k0", "v0"))
                .expectSegmentToBeOffloaded(broker0, topicA, p0, 1, new KeyValueSpec("k1", "v1"))
                .expectEarliestLocalOffsetInLogDirectory(topicA, p0, 2L)
                .produce(topicA, p0, new KeyValueSpec("k0", "v0"), new KeyValueSpec("k1", "v1"),
                        new KeyValueSpec("k2", "v2"))
                // new verification:
                 .expectListOffsets(topicA, 0, LatestTierSpec, 0(?)) <-- should create an expected `EpochEntry` instance
```
Let me know if you have any problem.

**Comment by showuon** [comment]
I mean, we can create a test called `listOffsetTest` under `org.apache.kafka.tiered.storage.integration` to test it.

**Comment by brandboat** [comment]
>Overall LGTM! We need integration tests for it. We should add some test like DeleteTopicTest using ExpectListOffsetsAction. I expected it should work like this (I didn't try):

Thanks @showuon , I enabled remote storage in GetOffsetShellTest, do we still need the test you mentioned above ?

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java [hunk 3]
```diff
@@ -416,7 +514,7 @@ private List<Row> expectedOffsetsForTopic(int i) {
 
     private List<Row> executeAndParse(String... args) {
         String out = ToolsTestUtils.captureStandardOut(() -> GetOffsetShell.mainNoExit(addBootstrapServer(args)));
-
+        System.out.println(out);
```
Comment (showuon) [inline tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: should be removed?

Comment (brandboat) [reply tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: yep, just removed.

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java [hunk 4]
```diff
@@ -70,31 +85,75 @@ private String getTopicName(int i) {
...
-                );
+                    .forEach(i -> IntStream.range(0, i * i)
+                            .forEach(msgCount -> assertDoesNotThrow(() -> producer.send(
+                                    new ProducerRecord<>(topicName.apply(i), msgCount % i, null, "val" + msgCount)).get())));
         }
     }
 
+    private static List<ClusterConfig> withRemoteStorage() {
+        Map<String, String> serverProperties = new HashMap<>();
+        serverProperties.put(RemoteLogManagerConfig.DEFAULT_REMOTE_LOG_METADATA_MANAGER_CONFIG_PREFIX + TopicBasedRemoteLogMetadataManagerConfig.REMOTE_LOG_METADATA_TOPIC_REPLICATION_FACTOR_PROP, "1");
```
Comment (showuon) [inline tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java:138]: Could we set `remote.log.metadata.topic.num.partitions` to 1 (default is 50) to speed up the test?

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java [hunk 5]
```diff
@@ -70,31 +85,75 @@ private String getTopicName(int i) {
...
+        serverProperties.put(RemoteLogManagerConfig.REMOTE_STORAGE_MANAGER_CLASS_NAME_PROP, LocalTieredStorage.class.getName());
+        serverProperties.put(RemoteLogManagerConfig.REMOTE_LOG_MANAGER_TASK_INTERVAL_MS_PROP, "1000");
+        serverProperties.put(ServerLogConfigs.LOG_CLEANUP_INTERVAL_MS_CONFIG, "1000");
+        serverProperties.put(ServerLogConfigs.LOG_INITIAL_TASK_DELAY_MS_CONFIG, "100");
+
+        Map<String, String> zkProperties = new HashMap<>(serverProperties);
+        zkProperties.put(RemoteLogManagerConfig.REMOTE_LOG_METADATA_MANAGER_LISTENER_NAME_PROP, "PLAINTEXT");
+
+        Map<String, String> raftProperties = new HashMap<>(serverProperties);
+        raftProperties.put(RemoteLogManagerConfig.REMOTE_LOG_METADATA_MANAGER_LISTENER_NAME_PROP, "EXTERNAL");
```
Comment (showuon) [inline tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: Could we add a comment here to explain why we need 2 different settings?

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java [hunk 6]
```diff
@@ -70,31 +85,75 @@ private String getTopicName(int i) {
...
+    private String getRemoteLogStorageEnabledTopicName(int i) {
+        return "topicRLS" + i;
+    }
+
     private void setUp() {
+        setupTopics(this::getTopicName, Collections.emptyMap());
+        sendProducerRecords(this::getTopicName);
+    }
+
+    private void setUpRemoteLogTopics() {
```
Comment (showuon) [inline tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java:97]: I'll add some comment for this method. Ex:
```
In this method, we'll create 4 topics and produce records to the log like this:
topicRLS1 -> 1 segment
topicRLS2 -> 2 segments (1 local log segment + 1 segment in the remote storage)
topicRLS3 -> 3 segments (1 local log segment + 2 segments in the remote storage)
topicRLS4 -> 4 segments (1 local log segment + 3 segments in the remote storage)
```

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java [hunk 7]
```diff
@@ -274,6 +333,45 @@ public void testGetOffsetsByMaxTimestamp() {
...
+            // unknown offsets are ignored by GetOffsetShell hence we have empty result here.
+            assertEquals(Collections.emptyList(),
+                    executeAndParse("--topic-partitions", "topic\\d+:0", "--time", time));
+
+            // test topics enable remote log storage
+            TestUtils.waitForCondition(() ->
+                    Arrays.asList(
+                            new Row("topicRLS2", 0, 0L),
+                            new Row("topicRLS3", 0, 1L),
+                            new Row("topicRLS4", 0, 2L))
```
Comment (showuon) [inline tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: I'll add 1 more line for the comment to explain why `topicRLS1` has no result:
`topicRLS1 has no result because there's no log segments being uploaded to the remote storage`

**Review by showuon — COMMENTED** [review]
Overall LGTM! Left some comments. Thanks.

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java (hunk 5 above)
Comment (brandboat) [reply tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: Correct me if I'm wrong, after spending some time digging through the code I found out that in Type.ZK the default listener name is `PLAINTEXT` and is set via
https://github.com/apache/kafka/blob/96989e4b6483e2e24883de20ff2cbcd2d3b39dbf/core/src/test/java/kafka/test/junit/ZkClusterInvocationContext.java#L312-L316
while in Type.KRAFT/Type.CO_KRAFT, it is `EXTERNAL`, see  https://github.com/apache/kafka/blob/1084d3b9c95aecccbe3c82e84ae4c8f406fc68e1/core/src/test/java/kafka/testkit/KafkaClusterTestKit.java#L324-L332

Another thing is that ClusterConfig#setListenerName only works under ZK mode, in KRAFT it is useless. Perhaps we should make it work under Raft mode too. gentle ping @chia7712, any thoughts ?

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java (hunk 6 above)
Comment (brandboat) [reply tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java:97]: Thank you for the detailed comment, this will certainly help others grasp the details of the test 😃

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java (hunk 7 above)
Comment (brandboat) [reply tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: Everything has been addressed in the latest commit, thank you

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java (hunk 5 above)
Comment (chia7712) [reply tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: @brandboat you are right. Could you please open a jira for that? `KRAFT` should honor the `listener name` and `security protocol` from ClusterConfig.

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java [hunk 8]
```diff
@@ -70,31 +85,80 @@ private String getTopicName(int i) {
...
+        serverProperties.put(RemoteLogManagerConfig.DEFAULT_REMOTE_LOG_METADATA_MANAGER_CONFIG_PREFIX + TopicBasedRemoteLogMetadataManagerConfig.REMOTE_LOG_METADATA_TOPIC_REPLICATION_FACTOR_PROP, "1");
+        serverProperties.put(RemoteLogManagerConfig.REMOTE_LOG_STORAGE_SYSTEM_ENABLE_PROP, "true");
+        serverProperties.put(RemoteLogManagerConfig.REMOTE_STORAGE_MANAGER_CLASS_NAME_PROP, LocalTieredStorage.class.getName());
+        serverProperties.put(RemoteLogManagerConfig.REMOTE_LOG_MANAGER_TASK_INTERVAL_MS_PROP, "1000");
+        serverProperties.put(TopicBasedRemoteLogMetadataManagerConfig.REMOTE_LOG_METADATA_TOPIC_PARTITIONS_PROP, "1");
+        serverProperties.put(ServerLogConfigs.LOG_CLEANUP_INTERVAL_MS_CONFIG, "1000");
+        serverProperties.put(ServerLogConfigs.LOG_INITIAL_TASK_DELAY_MS_CONFIG, "100");
+        serverProperties.put(RemoteLogManagerConfig.REMOTE_LOG_METADATA_MANAGER_LISTENER_NAME_PROP, "EXTERNAL");
+
+        return Arrays.asList(
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: We can set listener name for all types even though `setListenerName` is no-op for kraft/co-kraft mode. That can simplify the code and it is still valid.
```java
        return Collections.singletonList(
                ClusterConfig.defaultBuilder()
                        .setTypes(new HashSet<>(Arrays.asList(ZK, KRAFT, CO_KRAFT)))
                        .setServerProperties(serverProperties)
                        .setListenerName("EXTERNAL")
                        .build());
```

File: tools/src/main/java/org/apache/kafka/tools/GetOffsetShell.java [hunk 9]
```diff
@@ -283,6 +283,10 @@ private OffsetSpec parseOffsetSpec(String listOffsetsTimestamp) throws TerseExce
                 return OffsetSpec.latest();
             case "max-timestamp":
                 return OffsetSpec.maxTimestamp();
+            case "earliest-local":
```
Comment (chia7712) [inline tools/src/main/java/org/apache/kafka/tools/GetOffsetShell.java:287]: Could you please update docs?
https://github.com/apache/kafka/blob/2c9d7afe4c60656d10f5b15367e3fb65173d0372/tools/src/main/java/org/apache/kafka/tools/GetOffsetShell.java#L135

File: clients/src/main/java/org/apache/kafka/clients/admin/OffsetSpec.java [hunk 10]
```diff
@@ -73,18 +73,20 @@ public static OffsetSpec maxTimestamp() {
...
     public static OffsetSpec earliestLocalSpec() {
         return new EarliestLocalSpec();
     }
 
     /**
-     * Used to retrieve the offset with the highest offset of data stored in remote storage,
-     * and when Tiered Storage is not enabled, we won't return any offset (i.e. Unknown offset)
+     * Used to retrieve the highest offset of data stored in remote storage.
+     * <br/>
+     * Note: When tiered storage is not enabled, we will return unknown offset.
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/clients/admin/OffsetSpec.java:89]: Could you please add test for this scenario?

File: clients/src/main/java/org/apache/kafka/clients/admin/OffsetSpec.java [hunk 11]
```diff
@@ -73,18 +73,20 @@ public static OffsetSpec maxTimestamp() {
...
     /**
-     * Used to retrieve the offset with the local log start offset,
-     * log start offset is the offset of a log above which reads are guaranteed to be served
-     * from the disk of the leader broker, when Tiered Storage is not enabled, it behaves the same
-     * as the earliest timestamp
+     * Used to retrieve the local log start offset.
+     * Local log start offset is the offset of a log above which reads
+     * are guaranteed to be served from the disk of the leader broker.
+     * <br/>
+     * Note: When tiered Storage is not enabled, it behaves the same as retrieving the earliest timestamp offset.
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/clients/admin/OffsetSpec.java:80]: Could you please add test for it?

File: clients/src/main/java/org/apache/kafka/clients/admin/OffsetSpec.java [hunk 12]
```diff
@@ -73,18 +73,20 @@ public static OffsetSpec maxTimestamp() {
...
      */
     public static OffsetSpec earliestLocalSpec() {
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/clients/admin/OffsetSpec.java]: This is unrelated to this PR. The other specs don't use `Spec` as postfix. Maybe we should align the naming before release?

File: clients/src/main/java/org/apache/kafka/clients/admin/OffsetSpec.java [hunk 13]
```diff
@@ -73,18 +73,20 @@ public static OffsetSpec maxTimestamp() {
...
 
     /**
-     * Used to retrieve the offset with the highest offset of data stored in remote storage,
-     * and when Tiered Storage is not enabled, we won't return any offset (i.e. Unknown offset)
+     * Used to retrieve the highest offset of data stored in remote storage.
+     * <br/>
+     * Note: When tiered storage is not enabled, we will return unknown offset.
      */
     public static OffsetSpec latestTierSpec() {
         return new LatestTierSpec();
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/clients/admin/OffsetSpec.java]: This is not related to this PR, but KIP-1005 (https://cwiki.apache.org/confluence/display/KAFKA/KIP-1005%3A+Expose+EarliestLocalOffset+and+TieredOffset) calls it `LatestTieredSpec` rather than `LatestTierSpec`

ping @FrankYang0529 as you are the author of #16781

**Review by chia7712 — COMMENTED** [review]
@brandboat thanks for this nice patch. I leave some comments for this PR. The naming is part of public APIs, so we must have "accurate" naming before release.

File: clients/src/main/java/org/apache/kafka/clients/admin/OffsetSpec.java (hunk 13 above)
Comment (FrankYang0529) [reply clients/src/main/java/org/apache/kafka/clients/admin/OffsetSpec.java]: Thanks for finding this. Yes, we should follow the KIP. @brandboat, could you fix it in this PR? I'm afraid if I create another PR may block your progress.

File: clients/src/main/java/org/apache/kafka/clients/admin/OffsetSpec.java (hunk 12 above)
Comment (brandboat) [reply clients/src/main/java/org/apache/kafka/clients/admin/OffsetSpec.java]: OK, I'll remove the postfix

File: clients/src/main/java/org/apache/kafka/clients/admin/OffsetSpec.java (hunk 10 above)
Comment (brandboat) [reply clients/src/main/java/org/apache/kafka/clients/admin/OffsetSpec.java:89]: this is included in https://github.com/apache/kafka/pull/16783/files/b6426e51d40e5b921f7d497bb0f1c3b2b88adf68#diff-8c7debdc8479f762a5a9585de62d3254c10768249bddd45391b49ba5683526e4R363-R367

File: clients/src/main/java/org/apache/kafka/clients/admin/OffsetSpec.java (hunk 11 above)
Comment (brandboat) [reply clients/src/main/java/org/apache/kafka/clients/admin/OffsetSpec.java:80]: Sure, will do, thanks

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java (hunk 5 above)
Comment (brandboat) [reply tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: thanks for double checking, filed https://issues.apache.org/jira/browse/KAFKA-17256

File: clients/src/main/java/org/apache/kafka/clients/admin/OffsetSpec.java (hunk 13 above)
Comment (brandboat) [reply clients/src/main/java/org/apache/kafka/clients/admin/OffsetSpec.java]: Everything has been addressed, thanks everyone.

**Comment by FrankYang0529** [comment]
@brandboat, I think we also need to update L296-297 in GetOffsetShell

```
throw new TerseException("Malformed time argument " + listOffsetsTimestamp + ". " +
                            "Please use -1 or latest / -2 or earliest / -3 or max-timestamp, or a specified long format timestamp");
```

**Review by FrankYang0529 — APPROVED** [review]
LGTM. Thanks for the patch.

File: tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java [hunk 14]
```diff
@@ -70,31 +85,76 @@ private String getTopicName(int i) {
...
         }
     }
 
+    private static List<ClusterConfig> withRemoteStorage() {
+        Map<String, String> serverProperties = new HashMap<>();
+        serverProperties.put(RemoteLogManagerConfig.DEFAULT_REMOTE_LOG_METADATA_MANAGER_CONFIG_PREFIX + TopicBasedRemoteLogMetadataManagerConfig.REMOTE_LOG_METADATA_TOPIC_REPLICATION_FACTOR_PROP, "1");
+        serverProperties.put(RemoteLogManagerConfig.REMOTE_LOG_STORAGE_SYSTEM_ENABLE_PROP, "true");
+        serverProperties.put(RemoteLogManagerConfig.REMOTE_STORAGE_MANAGER_CLASS_NAME_PROP, LocalTieredStorage.class.getName());
+        serverProperties.put(RemoteLogManagerConfig.REMOTE_LOG_MANAGER_TASK_INTERVAL_MS_PROP, "1000");
+        serverProperties.put(TopicBasedRemoteLogMetadataManagerConfig.REMOTE_LOG_METADATA_TOPIC_PARTITIONS_PROP, "1");
```
Comment (showuon) [inline tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: I think we should add `RemoteLogManagerConfig.DEFAULT_REMOTE_LOG_METADATA_MANAGER_CONFIG_PREFIX` to set `REMOTE_LOG_METADATA_TOPIC_PARTITIONS_PROP`, but I might be wrong. Could you confirm the current change takes effect in the remote log topics?

Comment (brandboat) [reply tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java]: yeah... I missed that. Thank you

**Review by showuon — APPROVED** [review]
LGTM!

**Review by chia7712 — APPROVED** [review]
LGTM

**Comment by brandboat** [comment]
failed test passed in my local env:
```sh
./gradlew cleanTest \
  core:test --tests kafka.zk.ZkMigrationFailoverTest.testDriverSkipsEventsFromOlderEpoch \
  tool:test --tests org.apache.kafka.tools.LeaderElectionCommandTest.testPreferredReplicaElection \
  clients:test --tests org.apache.kafka.common.record.MemoryRecordsBuilderTest.testBuffersDereferencedOnClose
```
